### ISSUE TITLE
🐛 Bug Report: Sporadic JSON errors

### ISSUE THREAD

**Issue by cbmiller2610** [redlib-org/redlib#151]
## Describe the bug

Bug when clicking on post from r/popular

## Steps to reproduce the bug

1. Go to r/popular
2. Click on the post /r/Wellthatsucks/comments/1dnjif4/i_was_accepted_to_a_phd_program_4_years_ago_and_i

## What's the expected behavior?

  See the post

## Additional context / screenshot

 ![image](https://github.com/redlib-org/redlib/assets/7013753/e256ed21-ab28-4d71-971e-09758aff925d)

**Comment by freedit-dev** [comment]
Jun 02 13:28:37 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""quarantined"": "Forbidden"
Jun 02 13:28:49 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:29:15 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:29:35 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:29:46 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:29:48 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:29:50 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:30:03 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""quarantined"": "Forbidden"
Jun 02 13:30:20 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:30:25 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:30:30 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:30:41 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:30:42 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:30:46 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 "null": "Forbidden"
Jun 02 13:31:04 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:31:11 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:31:20 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Post ID is invalid. It may point to a post on a community that has been banned.
Jun 02 13:31:59 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""private"": "Forbidden"
Jun 02 13:32:24 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:33:19 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:34:16 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:34:22 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""private"": "Forbidden"
Jun 02 13:34:54 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:35:06 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:35:19 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:35:43 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:35:44 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:35:52 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:36:38 base redlib[67566]:  ERROR redlib::utils > Error page rendered: No posts found
Jun 02 13:36:39 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 404 ""banned"": "Not Found"
Jun 02 13:36:47 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:37:03 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:37:06 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""private"": "Forbidden"
Jun 02 13:37:54 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:38:23 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:38:42 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:39:12 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 "null": "Forbidden"
Jun 02 13:40:08 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:40:10 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:40:50 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:41:10 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:41:15 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Nothing here
Jun 02 13:41:34 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Reddit error 403 ""private"": "Forbidden"
Jun 02 13:42:05 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>
Jun 02 13:42:45 base redlib[67566]:  ERROR redlib::utils > Error page rendered: Couldn't send request to Reddit: error trying to connect: dns error: failed to lookup address information: Name or service not kno>

**Comment by freedit-dev** [comment]
Back to life when restarted.

**Comment by arch-btw** [comment]
Not sure if related but apparently not encoding "+" causes ratelimit errors:

https://sh.reddit.com/r/redditdev/comments/1diwa3i/anyone_getting_prawcoreexceptionsredirect/

**Comment by arulagrawal** [comment]
At least on my end, it appears the server I was using to host redlib has been blacklisted by reddit.

If I run `curl -L "https://reddit.com/r/AskReddit/comments/1dnet5b/which_real_life_cheat_codes_do_you_know/la39c
tt.json?&sort=confidence&raw_json=1"` from my personal home pc, I get the expected response, but on the VPS, I get a html response which contains the message "You've been blocked by network security." I'm guessing this is the same html that gets returned when you are blocked for using a VPN.

**Comment by urmum3728325481846239049372354837246382** [comment]
I am getting the same exact error when opening the homepage: Failed to parse page JSON data: EOF while parsing a value at line 1 column 0. is there a fix yet? BTW this happens on all the instances ive tried

**Comment by I-I-IT** [comment]
Some instances do work, like reddit.nerdvpn.de, and redlib.ducks.party

Rate limiting ?

**Comment by CaldeiraG** [comment]
I run a single-user instance and also getting the same errors so it seems to be something on reddit's side perhaps

**Comment by sigaloid** [comment]
I really can't tell what this problem is. I can't reproduce it locally, despite experiencing it all day on public instances.

Is anyone able to reproduce this consistently within a few minutes of starting a local server?

**Comment by sigaloid** [comment]
Hm, seems like there's something I could do to improve this (evicting invalid states from cache). Will put up a PR soon. 

I would love to get to the bottom of this, though, it requires adding a few println statements to see just precisely what the client is getting back.

**Comment by ButteredCats** [comment]
> I really can't tell what this problem is. I can't reproduce it locally, despite experiencing it all day on public instances.
> 
> Is anyone able to reproduce this consistently within a few minutes of starting a local server?
> 
> There's a tiny chance it could have been caused by this: #154 or maybe even this: [bacc9e3](https://github.com/redlib-org/redlib/commit/bacc9e35df7b8eb3a53b8335b3a53bbab88bf97a)
> 
> But I couldn't tell ya. Best thing to do is wait for instances to update to see

After 1-2 minutes even on 48873c01b9afb66aefbe1120a6637ba0fdfebd68 it still happens for me, but seemingly only on my server and not my desktop.

Server has two different containers running Redlib, each with a different public IPv6 address. All devices have the same public IPv4. Disabling IPv6 on my desktop still doesn't result in an error though.

I've got no problem taking one of the server containers offline to do some local testing and adding some print statements if you want to let me know where they'd go.

**Comment by freedit-dev** [comment]
`Status code: 429 Too Many Requests`

Reddit seems to be able to detect Redlib. Not by ip, even with my cloudflare proxy fork it still complains about too many requests.

**Comment by sigaloid** [comment]
Hm. I have to wonder if it's some kind of bot protection, since it's more likely to happen to public instances (and thus would be under load).

Thanks @ButteredCats, here's a diff, this should work on `main` right now.

```bash
mv logging.diff.txt logging.diff
git apply logging.diff
cargo run
```

[logging.diff.txt](https://github.com/user-attachments/files/15981036/logging.diff.txt)



Would greatly appreciate a zip with the `error_*` files that you get (note: this will include the ephemeral token generated on your device - would help me reproduce with the exact request, but not strictly necessary)

**Comment by sigaloid** [comment]
> Reddit seems to be able to detect Redlib. Not by ip, even with my cloudflare proxy fork it still complains about too many requests.

I think that means your IP is banned (including cloudflare). I can't reproduce a 429 off the bat at all

**Comment by freedit-dev** [comment]
@sigaloid My IP has been banned for a long time, at least several months. But a few days ago, my instance was still working normally.

**Comment by sigaloid** [comment]
Hm. What image are you using? I made a few theoretically relevant changes in the last few hours

**Comment by freedit-dev** [comment]
I'll try the latest commit.

**Comment by ButteredCats** [comment]
Couldn't recreate it with the container not getting proxied to, reenabled it for just a second to see if that'd cause it and was immediately hit with 429 errors.

However it only created an [error_oauth_token.txt](https://github.com/user-attachments/files/15981217/error_oauth_token.txt) and I was hit with a bunch of this:
```
thread 'tokio-runtime-worker' panicked at src/client.rs:355:85:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
 ERROR redlib::client > Got an invalid response from reddit EOF while parsing a value at line 1 column 0. Status code: 429 Too Many Requests
 ERROR redlib::client > Failed to parse JSON with content:
 ```

Unfortunately I have this awful thing called work tomorrow so I'm not gonna be able to test any more tonight.

**Comment by sigaloid** [comment]
Ah, irony of ironies. Couldn't test the logging because I couldn't reproduce it. Updated patch, no rush on reproduction.

**Comment by ButteredCats** [comment]
I procrastinated long enough to see that before I got off lol.

Hopefully this can tell you what you need to know: [errors.zip](https://github.com/user-attachments/files/15981313/errors.zip)

**Comment by sigaloid** [comment]
It does. Interestingly enough the error code is (presumably) still 200 (since it got all the way to json parsing). But the rate limit was reached (info in headers). Content-length is zero as well. 

I'm not sure what the solution path is. The goal was for oauth to remove the worry for rate limits. I need to probably add a lot of observability into the rate limiting info returned in the headers. For now, I think making sure the failures are evicted from cache instead of staying broken for 5 minutes is the stopgap solution. 

I think the reason you're dealing with rate limits could have something to do with IP reputation. Maybe some kind of temporary blocklist. Weird interplay with their own edge CDN and the way we present ourselves (totally legit mobile client) vs TLS cipher handshake appearing differently, plus datacenter IP etc. There's so many weird things they can do to detect us, it's not difficult to bypass, but narrowing exactly what changes every time is tedious. I guess this might mean they're actually trying to block us? 🤔

**Comment by arch-btw** [comment]
> datacenter IP

On VPN I've noticed that:

- One of the most popular VPN providers goes straight to the "You've been blocked page" **_almost_** always.
- With a lesser known VPN provider I can change servers a few times and get access without logging in.
- All of this happens less during evening/past midnight. I take it that there are less users using that specific IP address at that time.
- All of these are datacenter IP's, so I think what you're saying is correct. It's a combination of things adding up.

Funny enough every VPN server can bypass the message on **sh.reddit.com** at the moment, but this will likely be patched any second now.

**Comment by ButteredCats** [comment]
My server is in my home network, where I just have a normal residential plan.

When I have time I'll disable IPv6 on both my Redlib containers (which are still recieving requests) and my desktop and then see if my desktop starts running into the issue too.

**Comment by sigaloid** [comment]
Bad news, it looks like the issue lies in the fact that the rate limit started being enforced. Going to push a quick fix out that properly renders an error page, would greatly appreciate instance operators upgrading to it. 

Also, I'm going to close this issue - the error is going to be fixed. I'll open a new issue for the rate limiting issue.
